politeiawww/user.go

```diff
@@ -1110,6 +1113,112 @@ func (p *politeiawww) login(l www.Login) loginResult {
 		}
 	}
 
+	// First check if TOTP is enabled and verified.
+	if u.TOTPVerified {
+		// return error to alert that a code is required.
+		if l.Code == "" {
+			log.Debugf("login: totp code required %v", u.Email)
+			err = www.UserError{
+				ErrorCode: www.ErrorStatusRequiresTOTPCode,
+			}
+			return loginResult{
+				reply: nil,
+				err:   err,
+			}
+		}
+		requestTime := time.Now()
+		currentCode, err := p.totpGenerateCode(u.TOTPSecret, requestTime)
+		if err != nil {
+			return loginResult{
+				reply: nil,
+				err:   err,
+			}
+		}
+		// Check to see if provided code matches.
+		if currentCode != l.Code {
+			// Code did not match, is this the first failed attempt with
+			// this code?
+			if len(u.TOTPLastFailedCodeTime) == 0 {
+				log.Debugf("login: new totp code failure %v", u.Email)
+				u.TOTPLastFailedCodeTime = make([]int64, 0, 2)
+				u.TOTPLastFailedCodeTime = append(u.TOTPLastFailedCodeTime, requestTime.Unix())
+			} else {
+				// A code has already been generated, check to see if it
+				// matches the recently generated code.
+				oldCode, err := p.totpGenerateCode(u.TOTPSecret,
```

Comment:
I don't understand what is happening here.  We already know the code doesn't match.  What are we checking here?

Reply:
So the way the process works currently is if the "currentCode" does not match the code provided, there follows a series of checks to see if the user has already attempted to submit a TOTP token during the same period.  
If the code does not match and the user has no failed totp attempts seen, then just add the attempt to the list and return a fail.  If the code does not match and the user has totp failures we must check to see if the last failure seen is within the same "period" as before.  We take the last time appended to the failure list and then generate a code from that time, if the "oldCode" matches the "currentCode" then it's the same window and we just add another instance and return an error (if it's above the constant for failure attempts (2) then they get a Wait until next TOTP window error.  If the "oldCode" doesn't match the "currentCode" we know it's a new window and we create a new totp time failure list and the process can continue.

Reply:
@degeri care to review that logic to see if it looks ok?

Reply:
reviewed looks ok.

Reply:
@lukebp ok tried to clean up the code and add thorough comments to the logic paths are clear. 

I decided to go with some deeper nesting to allow for 1 user update and 1 replyError return at the bottom.  I can switch to less nesting but would need multiple user updates and returns, if you prefer.